Compile error with functor::power

The Cython `numpy_pythran_unit` test has fairly recently started failing with a C++ compile error. I'm not completely sure if it's a Cython or a Pythran issue, but downgrading Cython doesn't fix the problem, so I think something has changed with Pythran.

The code in question is (I've removed some irrelevant lines):

```
decltype(pythonic::numpy::functor::power{}(std::declval<pythonic::types::ndarray<double,pythonic::types::pshape<long>>>(), std::declval<long>())) __pyx_t_1;
new (&__pyx_t_1) decltype(__pyx_t_1){pythonic::numpy::functor::power{}(__pyx_v_values, ((long)3))};
```

`__pyx_v_values` is of type `pythonic::types::ndarray<double,pythonic::types::pshape<long>> __pyx_v_values`

The error message is

```
numpy_pythran_unit.cpp: In function ‘PyObject* __pyx_pf_18numpy_pythran_unit_10power(PyObject*, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >)’:
numpy_pythran_unit.cpp:6676:100: error: no matching function for call to ‘{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::power, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::broadcast<long int, long int> >::numpy_expr(<brace-enclosed initializer list>)’
 6676 |   new (&__pyx_t_1) decltype(__pyx_t_1){pythonic::numpy::functor::power{}(__pyx_v_values, ((long)3))};
      |                                                                                                    ^
In file included from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/ndarray.hpp:39,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/tuple.hpp:13,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:7,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from numpy_pythran_unit.cpp:805:
<home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/numpy_expr.hpp:38:3: note: candidate: ‘{anonymous}::pythonic::types::numpy_expr<O, Args>::numpy_expr(const Args& ...) [with Op = {anonymous}::pythonic::numpy::functor::power; Args = {{anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::broadcast<long int, long int>}]’
   38 |   numpy_expr<Op, Args...>::numpy_expr(Args const &... args)
      |   ^~~~~~~~~~~~~~~~~~~~~~~
<home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/numpy_expr.hpp:38:3: note:   candidate expects 2 arguments, 1 provided
In file included from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/ndarray.hpp:36,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/ndarray.hpp:4,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/tuple.hpp:13,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/builtins/bool_.hpp:7,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/NoneType.hpp:8,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/types/slice.hpp:5,
                 from <home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/core.hpp:29,
                 from numpy_pythran_unit.cpp:805:
<home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/numpy_expr.hpp:638:5: note: candidate: ‘constexpr {anonymous}::pythonic::types::numpy_expr<O, Args>::numpy_expr({anonymous}::pythonic::types::numpy_expr<O, Args>&&) [with Op = {anonymous}::pythonic::numpy::functor::power; Args = {{anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::broadcast<long int, long int>}]’
  638 |     numpy_expr(numpy_expr &&) = default;
      |     ^~~~~~~~~~
<home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/numpy_expr.hpp:638:16: note:   no known conversion for argument 1 from ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::power, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >&, {anonymous}::pythonic::types::broadcast<long int, long int> > >::type’ {aka ‘{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::power, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >&, {anonymous}::pythonic::types::broadcast<long int, long int> >’} to ‘{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::power, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::broadcast<long int, long int> >&&’
  638 |     numpy_expr(numpy_expr &&) = default;
      |                ^~~~~~~~~~~~~
<home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/numpy_expr.hpp:637:5: note: candidate: ‘{anonymous}::pythonic::types::numpy_expr<O, Args>::numpy_expr(const {anonymous}::pythonic::types::numpy_expr<O, Args>&) [with Op = {anonymous}::pythonic::numpy::functor::power; Args = {{anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::broadcast<long int, long int>}]’
  637 |     numpy_expr(numpy_expr const &) = default;
      |     ^~~~~~~~~~
<home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/numpy_expr.hpp:637:16: note:   no known conversion for argument 1 from ‘std::enable_if<true, {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::power, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >&, {anonymous}::pythonic::types::broadcast<long int, long int> > >::type’ {aka ‘{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::power, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >&, {anonymous}::pythonic::types::broadcast<long int, long int> >’} to ‘const {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::power, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::broadcast<long int, long int> >&’
  637 |     numpy_expr(numpy_expr const &) = default;
      |                ^~~~~~~~~~~~~~~~~~
<home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/numpy_expr.hpp:636:5: note: candidate: ‘constexpr {anonymous}::pythonic::types::numpy_expr<O, Args>::numpy_expr() [with Op = {anonymous}::pythonic::numpy::functor::power; Args = {{anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::broadcast<long int, long int>}]’
  636 |     numpy_expr() = default;
      |     ^~~~~~~~~~
<home dir>/.local/lib/python3.8/site-packages/pythran/pythonic/include/types/numpy_expr.hpp:636:5: note:   candidate expects 0 arguments, 1 provided
```

The Pythran version is 0.9.8.post2. Python version here is 3.8 but it fails fairly consistently across a lot of different Python versions.

--- a/pythran/pythonic/include/types/numpy_expr.hpp
+++ b/pythran/pythonic/include/types/numpy_expr.hpp
@@ -637,6 +637,14 @@ namespace types
     numpy_expr(numpy_expr const &) = default;
     numpy_expr(numpy_expr &&) = default;
 
+#ifdef CYTHON_ABI
+    template <class... Argp>
+    numpy_expr(numpy_expr<Op, Argp...> const &other)
+        : args(other.args)
+    {
+    }
+#endif
+
     numpy_expr(Args const &... args);
 
     template <size_t... I>

--- a/pythran/tests/cython/setup_add.py
+++ b/pythran/tests/cython/setup_add.py
@@ -1,5 +1,6 @@
 from distutils.core import setup
 from Cython.Build import cythonize
+import os
 
 setup(
     name = "add",

--- a/pythran/tests/test_cython.py
+++ b/pythran/tests/test_cython.py
@@ -1,33 +1,37 @@
 import os
+import glob
+import sys
 import unittest
 
 class TestCython(unittest.TestCase):
     pass
-# Needs to wait unil cython supports pythran new builtins naming
-#def add_test(name, runner, target):
-#    setattr(TestCython, "test_" + name, lambda s: runner(s, target))
-#
-#try:
-#    import Cython
-#    import glob
-#    import sys
-#    targets = glob.glob(os.path.join(os.path.dirname(__file__), "cython", "setup_*.py"))
-#    sys.path.append(os.path.join(os.path.dirname(__file__), "cython"))
-#
-#    for target in targets:
-#        def runner(self, target):
-#            cwd = os.getcwd()
-#            try:
-#                os.chdir(os.path.dirname(target))
-#                exec(open(os.path.basename(target)).read())
-#            except:
-#                raise
-#            finally:
-#                os.chdir(cwd)
-#        name, _ = os.path.splitext(os.path.basename(target))
-#        add_test(name, runner, target)
-#
-#
-#except ImportError:
-#    pass
+
+def add_test(name, runner, target):
+    setattr(TestCython, "test_" + name, lambda s: runner(s, target))
+
+for intermediate in glob.glob(os.path.join(os.path.dirname(__file__), "cython",
+                                           "*.cpp")):
+    os.remove(intermediate)
+
+try:
+    import Cython
+    targets = glob.glob(os.path.join(os.path.dirname(__file__), "cython", "setup_*.py"))
+    sys.path.append(os.path.join(os.path.dirname(__file__), "cython"))
+
+    for target in targets:
+        def runner(self, target):
+            cwd = os.getcwd()
+            try:
+                os.chdir(os.path.dirname(target))
+                exec(open(os.path.basename(target)).read())
+            except:
+                raise
+            finally:
+                os.chdir(cwd)
+        name, _ = os.path.splitext(os.path.basename(target))
+        add_test(name, runner, target)
+
+
+except ImportError:
+    pass
 

